### PR TITLE
Adding torch profiler

### LINKED ISSUE #10: Add GPU profiling infrastructure

### PR DESCRIPTION
**Pull request by amazloumi**
closes #10

- Adds torch profiler along with a scheduler to specify what training steps to profile:

for example the following config will skip the first step (step number 0) and warm up during step 1, 2, and 3. It will then profiles the step 4. it repeats this cycle only once. `repeat: 0` will continue the profiling cycle until the training loop completes.  
```
profiler:
  wait: 1
  warmup: 3
  active: 1
  repeat: 1
```

- Creates a directory named `profiler` under the `artifacts_path` directory to store the profiling outputs, a Json file for each profiling cycle.

### CODE REVIEW

**Comment by timothyngo** [comment]
> `repeat: 0` will continue the profiling cycle until the training loop completes.

I would expect `repeat:0` to repeat zero times, doing the cycle exactly once. Can you set the "until the training loop completes" setting to be something like `repeat: None` or `repeat:-1`?

**Comment by timothyngo** [comment]
What does it mean for the profiler to warm up?

**Comment by amazloumi** [comment]
> What does it mean for the profiler to warm up?

Usually we should not to start profiling from the very beginning of training due to some warmup overheads like memory allocations, initialization and kernel compilation. Using scheduler you can let the torch.profiler know that no need to record profiling during that warm up iterations.

**Comment by amazloumi** [comment]
> > `repeat: 0` will continue the profiling cycle until the training loop completes.
> 
> I would expect `repeat:0` to repeat zero times, doing the cycle exactly once. Can you set the "until the training loop completes" setting to be something like `repeat: None` or `repeat:-1`?

This is torch profiler config and how it works: https://github.com/pytorch/pytorch/blob/main/torch/profiler/profiler.py#L450:L451

We can make our own logic and have wrapper for the torch config that I dont think to be a good idea.

**Review by timothyngo — APPROVED** [review]
Okay, thanks for explaining! Seems good to me.

**Review by nikhilanand91 — APPROVED** [review]
